src/guards/role.guard.ts

```diff
@@ -0,0 +1,28 @@
+import { CanActivate, ExecutionContext, Injectable, UnauthorizedException } from "@nestjs/common";
+import { Reflector } from "@nestjs/core";
+import { Role } from "../user/entities/role.enum";
+import { ROLES_KEY } from "../decorators/has-roles.decorator";
+
+@Injectable()
+export class RolesGuard implements CanActivate{
+    constructor (private reflector: Reflector){}
+
+    canActivate(context: ExecutionContext): boolean {
+        const requiredRoles = this.reflector.getAllAndOverride<Role[]>(ROLES_KEY, [
+            context.getHandler(),
+            context.getClass()
+        ]);
+
+        if (!requiredRoles) {
+            return true;
+          }
+          const { user } = context.switchToHttp().getRequest();
+          console.log(user);
+          const hasRequiredRole = requiredRoles.some((role) => user?.role?.includes(role));
+        
+          if(!hasRequiredRole){
+            throw new UnauthorizedException('You need an admin role to perform this action.');
```

Comment:
Update this message to tell the user that they don't have the required permission to execute the action, instead of telling them they're not admin users